### ISSUE TITLE
Postgres vacuum prevented

### ISSUE THREAD

**Issue by jimfulton** [zodb/relstorage#147]
In #87 there's a mention of connections being left in the "idle in transaction" state which prevents autovacuum from working. That led me to:

https://blog.gocept.com/2012/05/22/dont-stop-postgresqls-autovacuum-with-your-application/

I'm interested because in looking at insert performance:

  https://github.com/jimfulton/zodb-jsonb-experiment#insert-performance (Search for vacuum)

It looks like the object_state table wasn't getting vacuumed, presumably because of the idle transaction. Manually doing a full vacuum (because this was a staging DB and I could get an exclusive lock) increased performance 40x.

This appears to be a huge problem.

It seems we should do a commit (or close) when we close ZODB connections. I'd be happy to add a ZODB hook for that. This would address the common case of web processes that close connections when finished.

In the longer term, we could add a ZODB mode that requires explicit transaction begins.

**Comment by jimfulton** [comment]
Although as long as all of the clients get restarted at some point, vacuuming up to that point should work.  I suppose the database I was looking at had clients running for a very long time.  I happened to have restarted them recently, before I realized this was an issue.

**Comment by jamadden** [comment]
Based on Shane's comments, I just assumed that #87 would have alleviated these problems, but I didn't actually check. 

Let me see if I understand the situation now:

When the storage has `tpc_finish` called on it, it goes to `_finish` where it explicitly rolls back the load connection. Then it commits the store connection, followed by releasing the commit lock (in the store connection). Last, it tells the cache that tpc_finish is done (which doesn't interact with the DB at all).

In postgres, this amounts to:
```python
load_connection.rollback() # rollback load 
store_connection.commit() # commit phase 2
pass # no-op to release the lock, released by commit
pass # no-op to cache
```

So after `tpc_finish`, it's not clear to me how we could have a connection idle in a transaction.

The next time `load` is called, the storage will notice that the load connection has been rolled back and will open it and perform a poll. At that point we should be idle in a transaction.

If no persistent objects are added or updated, and thus the Connection never joins the transaction, then the storage's tpc_finish will never be called, and we will remain idle in a transaction for the load connection.

That is, if a particular transaction is read-only, this state can arise until the next time a transaction is commited that has written something?

Having the Connection alert the storage when it (the connection) is returned to the DB pool (Connection.close) would allow the storage to take action at that point. Interestingly, rolling back the load connection at that point would be a subtle change of semantics: the next `load` would restart the poll cycle, jumping our MVCC view forward; currently that only happens when a Connection/storage write an object.

**Comment by jimfulton** [comment]
On Thu, Dec 15, 2016 at 10:48 AM, Jason Madden <notifications@github.com>
wrote:

> Based on Shane's comments, I just assumed that #87
> <https://github.com/zodb/relstorage/issues/87> would have alleviated
> these problems, but I didn't actually check.
>
An easy way to check is to:

  select * from pg_stat_activity;

And look for connections in the idle in transaction state.

Let me see if I understand the situation now:
>
> When the storage has tpc_finish called on it, it goes to _finish where it
> explicitly rolls back the load connection. Then it commits the store
> connection, followed by releasing the commit lock (in the store
> connection). Last, it tells the cache that tpc_finish is done (which
> doesn't interact with the DB at all).
>
> In postgres, this amounts to:
>
> load_connection.rollback() # rollback load
> store_connection.commit() # commit phase 2pass # no-op to release the lock, released by commitpass # no-op to cache
>
> So after tpc_finish, it's not clear to me how we could have a connection
> idle in a transaction.
>

Currently, in ZODB, transaction boundaries are implicit.
Connection.afterCompletion calls Connection.newTransaction, which calls
poll_invalidations, which starts a new Postgres transaction.

Also, we don't do any tpc dance for transactions that don't write.


> The next time load is called, the storage will notice that the load
> connection has been rolled back and will open it and perform a poll. At
> that point we should be idle in a transaction.
>
I don't think that's how it works. But perhaps I'm mistaken, but see above
(for ZODB 5).


> If no persistent objects are added or updated, and thus the Connection
> never joins the transaction, then the storage's tpc_finish will never be
> called, and we will remain idle in a transaction for the load connection.
>
> That is, if a particular transaction is read-only, this state can arise
> until the next time a transaction is commited that has written something?
>
Yes, see above.



> Having the Connection alert the storage when it (the connection) is
> returned to the DB pool (Connection.close) would allow the storage to take
> action at that point.
>
Yes.

> Interestingly, rolling back the load connection at that point would be a
> subtle change of semantics: the next load would restart the poll cycle,
> jumping our MVCC view forward;
>
I though that in ZODB 5, this would happen when poll_invalidations is
called.  That's what it certainly looks like is happening in the
poll_invalidations code on master.


> currently that only happens when a Connection/storage write an object.
>
Huh?

-- 
Jim Fulton
http://jimfulton.info

**Comment by jamadden** [comment]
Gah, I *always* forget about afterCompletion. So yes, that calls `Connection.newTransaction` which calls `storage.poll_invalidations` which restarts the load connection/transaction. If the Connection is then closed, that transaction sits idle without being used.

`Connection.open` also calls `Connection.newTransaction`. Beginning a transaction also calls `newTransaction` for all registered synchronizers.

The comments in afterCompletion indicate that it is there mostly as an optimization and for compatibility with non-hygeinic applications. Long-term, I wouldn't mind seeing that go away so that transaction boundaries only start when actively using intending for them to do so. But I know that's probably a big change.

(It does seem like there's potentially a lot of extra polling/synching going on, though. Commiting a transaction calls afterCompletion calls newTransaction calls poll_invalidations. The user then begins a new transaction, which calls newTransaction calls poll_invalidations. Or the transaction is commited which calls afterCompletion calls newTransaction calls poll_invalidation. The Connection is then closed and put at the front of the pool, where as soon at is it opened again newTransaction calls poll_invalidations.)

>> The next time load is called, the storage will notice that the load
>> connection has been rolled back and will open it and perform a poll. At
>> that point we should be idle in a transaction.
> I don't think that's how it works. But perhaps I'm mistaken, but see above
> (for ZODB 5).

From a pure storage perspective, RelStorage assumes it can be in that state. Connection may not use it that way, but the storage code supports it. 

In the meantime, knowing that Connection.open calls newTransaction calls poll_invalidations, and knowing that the storage itself will handle other uses, a callback to let the storage know when the connection is being closed and it can rollback its RDBMS connections seems fine to me. 

RelStorage today could just rollback the load connection after `poll_invalidations` and it would notice that and open it again when it was next used to `load` something, but there's a window for a race condition there where the Connection's PickleCache gets out of sync. If a storage could *push* invalidations up to the connection, then that could be solved

**Comment by jimfulton** [comment]
On Thu, Dec 15, 2016 at 11:54 AM, Jason Madden <notifications@github.com>
wrote:

> Gah, I *always* forget about afterCompletion. So yes, that calls
> Connection.newTransaction which calls storage.poll_invalidations which
> restarts the load connection/transaction. If the Connection is then closed,
> that transaction sits idle without being used.
>
> Connection.open also calls Connection.newTransaction. Beginning a
> transaction also calls newTransaction for all registered synchronizers.
>
> The comments in afterCompletion indicate that it is there mostly as an
> optimization and for compatibility with non-hygeinic applications.
> Long-term, I wouldn't mind seeing that go away so that transaction
> boundaries only start when actively using intending for them to do so. But
> I know that's probably a big change.
>
We discussed this at length somewhere :) a few months ago.  I don't
remember all of the details. This resulted in me posting
https://groups.google.com/forum/#!topic/python-transaction/dPcjNoKR_q4

> (It does seem like there's potentially a lot of extra polling/synching
> going on, though. Commiting a transaction calls afterCompletion calls
> newTransaction calls poll_invalidations. The user then begins a new
> transaction, which calls newTransaction calls poll_invalidations. Or the
> transaction is commited which calls afterCompletion calls newTransaction
> calls poll_invalidation. The Connection is then closed and put at the front
> of the pool, where as soon at is it opened again newTransaction calls
> poll_invalidations.)
>
Yup.  Unfortunately, given current implicit transaction boundaries, I don't
see a way around this.

In response to some use cases mentioned here:
https://github.com/zopefoundation/transaction/issues/35

I've been thinking of starting a discussion on the transaction list to
optionally make transaction boundaries explicit.  If transactions were
explicit, then afterCompletion wouldn't have to call newTransaction.

Related, we could add an explicit-transaction option to DB that would
simply cause Connection.afterCompletion to stop calling newTransaction.
That wouldn't require a change to the transaction package.

The most common use case, web-request-based transactions could benefit from
this without affecting applications.


> The next time load is called, the storage will notice that the load
> connection has been rolled back and will open it and perform a poll. At
> that point we should be idle in a transaction.
> I don't think that's how it works. But perhaps I'm mistaken, but see above
> (for ZODB 5).
>
> From a pure storage perspective, RelStorage assumes it can be in that
> state. Connection may not use it that way, but the storage code supports it.
>
> In the meantime, knowing that Connection.open calls newTransaction calls
> poll_invalidations, and knowing that the storage itself will handle other
> uses, a callback to let the storage know when the connection is being
> closed and it can rollback its RDBMS connections seems fine to me.
>
I don't like the idea of tying this to close.  I'd rather use an API like
afterCompletion, which ZODB would call on close, but would also call in
afterCompletion if and when we make transaction boundaries explicit.



> RelStorage today could just rollback the load connection after
> poll_invalidations and it would notice that and open it again when it was
> next used to load something, but there's a window for a race condition
> there where the Connection's PickleCache gets out of sync. If a storage
> could *push* invalidations up to the connection, then that could be solved
>
It seems a core tenet of RelStorage that it relies on the underlying RDBMS
to provide MVCC. Rolling back after poll would break that.

I don't want to treat this problem with bandaids.  From a practical point
of view, I think this problem can be mostly mitigated by calling a new
afterCompletion storage hook from afterCompletion and arrange that this is
also called on close.

So, in summary, I propose to


   1. add an afterCompletion (feel free to suggest another name) method to
   IMVCCStorage (or a subinterface) that's called from
   Connection.afterCompletion and when a connection is closed.
   2. add an explicit-transaction option to DB (and the zodb config
   element) that stops calling newTransaction in afterCompletion.
      1. I think we should then add seatbelts to RelStorage and the
      mvccadapter to error if most operations were invoked between a call to
      afterCompletion and newTransaction.


Thoughts?

-- 
Jim Fulton
http://jimfulton.info

**Comment by jamadden** [comment]
> It seems a core tenet of RelStorage that it relies on the underlying RDBMS
> to provide MVCC. Rolling back after poll would break that.

That's the race condition I was talking about.

>    1. add an afterCompletion (feel free to suggest another name) method to
>   IMVCCStorage (or a subinterface) that's called from
>   Connection.afterCompletion and when a connection is closed.
`on_transaction_idle`?

So presumably in RelStorage this function would rollback the load connection after we poll, otherwise it doesn't solve the problem.

But if this is called from `Connection.afterCompletion`, then the currently working "implicit" transaction support is broken by the same race condition we were talking about above, isn't it? I debugged through some examples and can easily produce cases that don't call `newTransaction`:

```python
conn = db.open() # calls newTransaction
root = conn.root()
transaction.commit() 
# That called afterCompletion/newTransaction. Imagine that we rolled back the load connection there
root['a thing'] = 'a value'
# That called conn.register, which joined a transaction via self.transaction_manager.get().join(self)
# which *does not* result in a call to newTransaction. At some point in the future, 
# if Connection.get/storage.load
# is called, the load connection will get restarted at some point in time
# ahead of what the picklecache has valid data for.
```

That's why I was suggesting this be done only from `Connection.close`, because then we're guaranteed to call `newTransaction` again (from open()) before using the Connection.

But maybe I'm missing something? Are you suggesting that this new hook only be called when the DB is in explicit-transaction mode?

Maybe the problem is that `TransactionManager.get()`  [bypasses calling the synchronizers even when it begins a new transaction](https://github.com/zopefoundation/transaction/blob/master/transaction/_manager.py#L77). I have a vague memory of that being discussed somewhere before.

**Comment by jimfulton** [comment]
On Thu, Dec 15, 2016 at 3:49 PM, Jason Madden <notifications@github.com>
wrote:

> It seems a core tenet of RelStorage that it relies on the underlying RDBMS
> to provide MVCC. Rolling back after poll would break that.
>
> That's the race condition I was talking about.
>

I figured.


>
>    1. add an afterCompletion (feel free to suggest another name) method to
>    IMVCCStorage (or a subinterface) that's called from
>    Connection.afterCompletion and when a connection is closed.
>    on_transaction_idle?
>
> I wonder where that last line came from.

>
>
> So presumably in RelStorage this function would rollback the load
> connection after we poll, otherwise it doesn't solve the problem.
>
No. It would only rollback the load connection (as you do now in tpc_finish
-- you could stop doing it there).  Then newTransaction would call
poll_invalidations as now.



> But if this is called from Connection.afterCompletion, then the currently
> working "implicit" transaction support is broken by the same race condition
> we were talking about above, isn't it? I debugged through some examples and
> can easily produce cases that don't call newTransaction:
>
> conn = db.open() # calls newTransaction
> root = conn.root()
> transaction.commit() # That called afterCompletion/newTransaction. Imagine that we rolled back the load connection there
> root['a thing'] = 'a value'# That called conn.register, which joined a transaction via self.transaction_manager.get().join(self)# which *does not* result in a call to newTransaction. At some point in the future, # if Connection.get/storage.load# is called, the load connection will get restarted at some point in time# ahead of what the picklecache has valid data for.
>
> That's why I was suggesting this be done only from Connection.close, because then we're guaranteed to call newTransaction again (from open()) before using the Connection.
>
> But maybe I'm missing something?
>
I think so. :)

> Are you suggesting that this new hook only be called when the DB is in
> explicit-transaction mode?
>
>
Let me try to spell this out blow by blow. First the implicit case.


   1. A transaction ends (commit or abort)
      1. connection.afterCompletion is called.
         1. It calls afterCompletion on the storage
         2. It calls newTransaction
            1. It calls poll invalidations
         2. Things happen :)
   3. Transaction ends go to step 1

Note that nothing rolls back after poll_invalidations until the transaction
ends and we start over.

On connection close, close calls afterCompletion on the storage because it
knows the current transaction is done as far as it's concerned.

In the explicit transaction case, a begin follows an end and we have the
extra newTransaction/poll_invalidation call. Bummer.

So we add the explicit-transaction DB option and we no longer make the
extra newTransaction/poll_invalidations call. (For safety, we also error if
an attempt is made to read or write without a newTransaction call.

Does that make sense?

If not, we should probably discuss it on gitter or slack.

Maybe the problem is that TransactionManager.get() bypasses calling the
> synchronizers even when it begins a new transaction
> <https://github.com/zopefoundation/transaction/blob/master/transaction/_manager.py#L77>.
> I have a vague memory of that being discussed somewhere before.
>

Yes, and it's coming back to me. :)

get isn't called at the start of the transaction. It's only the start as
far as the TM is concerned. The DM may have started the transaction much
sooner.

I decided to document this:
https://github.com/zopefoundation/transaction/pull/40

get creates a transaction when it needs to. This is a small optimization to
avoid allocating a resource that  might never be used.

Another thought I had was to add an explicit transaction mode. In that
mode, get would error if called without a preceding begin call.  But that's
a lower priority for me given that we can add a seatbelt in ZODB.

**Comment by jamadden** [comment]
> Does that make sense?

Yes, I see what I had confused now. For one thing, I had the order of steps 1 and 2 backwards. For another thing, I was treating the changes to `Connection.afterCompletion` is isolation of the changes to `Connection.close`. They need to happen together (or at least, `Connection.close` changes *still* have to happen) to solve the problem.

Thanks for the detailed explanation.

> I decided to document this:
> zopefoundation/transaction#40

> Another thought I had was to add an explicit transaction mode. In that
> mode, get would error if called without a preceding begin call. 

I had that same thought...

**Comment by jimfulton** [comment]
OK. So I'm going to:

- Update Connection to call afterCompletion, if present, on the storage as described above.

- Add an explicit-transaction option to DB.  I'll add the seatbelts in Connection so you don't have to in RelStorage.

Later we can talk about adding an explicit transaction mode on the transaction list.

**Comment by jimfulton** [comment]
I'll also make a RelStorage PR to add the afterCompletion method. I can tests this with PG to make sure these changes actually fix things. :)

**Comment by jamadden** [comment]
Regarding timeframe: Python 3.6 is expected very soon (at one point it was expected tomorrow, that may have slipped a bit due to a last-minute dict patch, but I don't know). Given that RelStorage 2.0rc1 went out on Monday, I was expecting to do a 2.0 final just after 3.6 becomes available, probably next Monday. However, given that I had hoped this was fixed already, it seems reasonable to include these changes in a 2.0rc2 if you're planning to get to them soon-ish. If not, they can go in the first 2.0 bugfix.

**Comment by jimfulton** [comment]
I wouldn't hold 2.0 for this.  This issue has existed for a long time.

**Comment by jimfulton** [comment]
I just found another reason to fix this. New indexes don't get used is there are any clients have been idle in transaction since before the index was created. Gaaaa.

I need to fix this soon.

**Comment by jimfulton** [comment]
FTR, I'm working on this now. ZODB PR soon.

**Comment by jimfulton** [comment]
https://github.com/zopefoundation/ZODB/pull/145

**Comment by jmuchemb** [comment]
> The comments in afterCompletion indicate that it is there mostly as an optimization and for compatibility with non-hygeinic applications. Long-term, I wouldn't mind seeing that go away so that transaction boundaries only start when actively using intending for them to do so. But I know that's probably a big change.

In the past, I thought the same thing and that's one of the reason I committed Nexedi/erp5@8995b64508f6d7c0fee1ef475de3dc3145ddc84d in ERP5.

IOW, _applications_ include unit tests, which are naturally written to start transaction implicitly, without issue. I don't think ZODB tests are ready to run in this mode. How do you plan to run them for RelStorage (I didn't check if you reuse them, as we do in NEO) ?

**Comment by jimfulton** [comment]
See https://github.com/zopefoundation/transaction/pull/43

Explicit mode is optional, so existing tests should be unaffected.

ZODB will continue the current behavior in non-explicit mode.

**Comment by jimfulton** [comment]
Smaller ZODB PR:

https://github.com/zopefoundation/ZODB/pull/146

**Comment by jimfulton** [comment]
Sharing some experience.

Getting transactions rolled back on close if a win for web application frameworks that close connections at the end of a request. Yay.

But in a legacy pyramid (:-)) app I'm working on, I'm having to deal with daemon processes that tend to keep connections open more or less forever.  Sometimes, these frameworks are so helpful, it's hard to change the connection management (or impractical for legacy apps).

I just resorted to:

    transaction.manager.explicit = True
    root._p_jar.explicit_transactions = True

(and added a ``begin()`` call.)

It's not advertized that you can change the ``explicit`` setting after construction. I think it should be.  Doing so makes it much easier to deal with the standard manager.

I'm thinking that connection.expliciit_transactions should be replaced with a query of the transaction managers's ``explicit`` attribute.

**Comment by jimfulton** [comment]
Note that this also affects mysql with an innodb back end. From:

https://dev.mysql.com/doc/refman/5.7/en/innodb-multi-versioning.html

"Commit your transactions regularly, including those transactions that issue only consistent reads. Otherwise, InnoDB cannot discard data from the update undo logs, and the rollback segment may grow too big, filling up your tablespace."

I bet $.05 that this is an issue for Oracle as well. I suggest adding a note about this to the RelStorage docs. I'll create an issue.
